orttraining/orttraining/training_api/optimizer.cc

```diff
@@ -80,16 +82,9 @@ Status Optimizer::ConstructInputs() {
   return Status::OK();
 }
 
-Optimizer::Optimizer(const std::string& optim_path_or_bytes,
-                     const std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters) : named_parameters_(parameters) {
-  // TODO: share threadpool, env with module session
-  const SessionOptions session_options;
-  std::unique_ptr<Environment> env;
-  ORT_ENFORCE(Environment::Create(nullptr, env) == Status::OK(), "Enviroment creation fails.");
-  optim_sess_ = std::move(std::make_unique<InferenceSession>(session_options, *env));
-
-  ORT_THROW_IF_ERROR(optim_sess_->Load(optim_path_or_bytes));
-  ORT_THROW_IF_ERROR(optim_sess_->Initialize());
+Optimizer::Optimizer(const std::unordered_map<std::string, std::shared_ptr<Parameter>>& named_parameters,
+                     InferenceSession* optim_session) : named_parameters_(named_parameters) {
```

Comment:
We should not use references here since we plan to hold onto that address for the lifetime of this class instance. We could transfer ownership of the unique pointer if the user needs access to the optim_session creation.
But the way I see it, the user doesn't care about the session creation as much as long as the session options and env are applied correctly. So, optimizer and module class can take the responsibility of creating their respective sessions and store them as unique pointers within the class instance.

Reply:
Have reverted the change passing Session pointer back to create InferenceSession inside Module.